File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/TabLayoutSelectionsObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.design.widget.TabLayout;
import android.support.design.widget.TabLayout.Tab;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class TabLayoutSelectionsObservable extends Observable<Tab> {
  private final TabLayout view;

  TabLayoutSelectionsObservable(TabLayout view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Tab> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnTabSelectedListener(listener);
    int index = view.getSelectedTabPosition();
    if (index != -1) {
      observer.onNext(view.getTabAt(index));
    }
  }

  static final class Listener extends MainThreadDisposable implements TabLayout.OnTabSelectedListener {
    private final TabLayout tabLayout;
    private final Observer<? super Tab> observer;

    Listener(TabLayout tabLayout, Observer<? super Tab> observer) {
      this.tabLayout = tabLayout;
      this.observer = observer;
    }

    @Override protected void onDispose() {
      tabLayout.removeOnTabSelectedListener(this);
    }

    @Override public void onTabSelected(Tab tab) {
      if (!isDisposed()) {
        observer.onNext(tab);
      }
    }

    @Override public void onTabUnselected(Tab tab) {
    }

    @Override public void onTabReselected(Tab tab) {
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/TabLayoutSelectionsObservable.kt
@file:JvmName("RxTabLayout")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import androidx.annotation.CheckResult
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.BaseOnTabSelectedListener
import com.google.android.material.tabs.TabLayout.Tab
import com.jakewharton.rxbinding3.internal.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable which emits the selected tab in `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* If a tab is already selected, it will be emitted immediately on subscribe.
 */
@CheckResult
fun TabLayout.selections(): Observable<Tab> {
  return TabLayoutSelectionsObservable(this)
}

private class TabLayoutSelectionsObservable(
  private val view: TabLayout
) : Observable<Tab>() {

  override fun subscribeActual(observer: Observer<in Tab>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view,
        observer)
    observer.onSubscribe(listener)
    view.addOnTabSelectedListener(listener)
    val index = view.selectedTabPosition
    if (index != -1) {
      observer.onNext(view.getTabAt(index)!!)
    }
  }

  private class Listener(
    private val tabLayout: TabLayout,
    private val observer: Observer<in Tab>
  ) : MainThreadDisposable(), BaseOnTabSelectedListener<Tab> {

    override fun onDispose() {
      tabLayout.removeOnTabSelectedListener(this)
    }

    override fun onTabSelected(tab: Tab) {
      if (!isDisposed) {
        observer.onNext(tab)
      }
    }

    override fun onTabUnselected(tab: Tab) {}

    override fun onTabReselected(tab: Tab) {}
  }
}
